File: src/main/java/no/nav/sosialhjelp/soknad/business/service/systemdata/FamilieSystemdata.java
package no.nav.sosialhjelp.soknad.business.service.systemdata;

import no.nav.sbl.soknadsosialhjelp.soknad.JsonData;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKildeSystem;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonNavn;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonAnsvar;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonBarn;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonEktefelle;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonErFolkeregistrertSammen;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonFamilie;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonForsorgerplikt;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonHarForsorgerplikt;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonSivilstatus;
import no.nav.sosialhjelp.soknad.business.service.soknadservice.Systemdata;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.person.PersonService;
import no.nav.sosialhjelp.soknad.person.domain.Barn;
import no.nav.sosialhjelp.soknad.person.domain.Ektefelle;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonSivilstatus.Status.GIFT;
import static org.apache.commons.lang3.StringUtils.isEmpty;

@Component
public class FamilieSystemdata implements Systemdata {

    private final PersonService personService;

    public FamilieSystemdata(PersonService personService) {
        this.personService = personService;
    }

    @Override
    public void updateSystemdataIn(SoknadUnderArbeid soknadUnderArbeid, String token) {
        JsonData jsonData = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData();
        String personIdentifikator = jsonData.getPersonalia().getPersonIdentifikator().getVerdi();
        JsonFamilie familie = jsonData.getFamilie();
        JsonSivilstatus systemverdiSivilstatus = innhentSystemverdiSivilstatus(personIdentifikator);

        if (systemverdiSivilstatus != null || familie.getSivilstatus() == null || familie.getSivilstatus().getKilde().equals(JsonKilde.SYSTEM)) {
            familie.setSivilstatus(systemverdiSivilstatus);
        }

        JsonForsorgerplikt forsorgerplikt = familie.getForsorgerplikt();
        JsonHarForsorgerplikt harForsorgerplikt = forsorgerplikt.getHarForsorgerplikt();
        JsonForsorgerplikt systemverdiForsorgerplikt = innhentSystemverdiForsorgerplikt(personIdentifikator);

        if (systemverdiForsorgerplikt.getHarForsorgerplikt().getVerdi()) {
            forsorgerplikt.setHarForsorgerplikt(systemverdiForsorgerplikt.getHarForsorgerplikt());

            List<JsonAnsvar> ansvarList = forsorgerplikt.getAnsvar();
            if (ansvarList != null && !ansvarList.isEmpty()) {
                ansvarList.removeIf(jsonAnsvar -> jsonAnsvar.getBarn().getKilde().equals(JsonKilde.SYSTEM) &&
                        isNotInList(jsonAnsvar, systemverdiForsorgerplikt.getAnsvar()));
                ansvarList.addAll(systemverdiForsorgerplikt.getAnsvar().stream()
                        .filter(sysAnsvar -> isNotInList(sysAnsvar, forsorgerplikt.getAnsvar()))
                        .collect(Collectors.toList()));
            } else {
                forsorgerplikt.setAnsvar(systemverdiForsorgerplikt.getAnsvar());
            }

        } else if (harForsorgerplikt == null || harForsorgerplikt.getKilde().equals(JsonKilde.SYSTEM) || !harForsorgerplikt.getVerdi()) {
            forsorgerplikt.setHarForsorgerplikt(systemverdiForsorgerplikt.getHarForsorgerplikt());
            forsorgerplikt.setBarnebidrag(null);
            forsorgerplikt.setAnsvar(new ArrayList<>());
        }
    }

    private boolean isNotInList(JsonAnsvar jsonAnsvar, List<JsonAnsvar> jsonAnsvarList) {
        return jsonAnsvarList.stream().noneMatch(
                ansvar -> {
                    if (ansvar.getBarn() == null) {
                        throw new IllegalStateException("JsonAnsvar mangler barn. Ikke mulig å skille fra andre barn");
                    }
                    if (ansvar.getBarn().getPersonIdentifikator() != null) {
                        return ansvar.getBarn().getPersonIdentifikator().equals(jsonAnsvar.getBarn().getPersonIdentifikator());
                    } else {
                        return ansvar.getBarn().getNavn().equals(jsonAnsvar.getBarn().getNavn());
                    }
                }
        );
    }

    private JsonSivilstatus innhentSystemverdiSivilstatus(String personIdentifikator) {
        var person = personService.hentPerson(personIdentifikator);
        if (person == null || isEmpty(person.getSivilstatus())) {
            return null;
        }

        Ektefelle ektefelle = person.getEktefelle();
        JsonSivilstatus.Status status = JsonSivilstatus.Status.fromValue(person.getSivilstatus());
        if (!GIFT.equals(status) || ektefelle == null) {
            return null;
        }

        return new JsonSivilstatus()
                .withKilde(JsonKilde.SYSTEM)
                .withStatus(status)
                .withEktefelle(tilSystemregistrertJsonEktefelle(ektefelle))
                .withEktefelleHarDiskresjonskode(ektefelle.getIkkeTilgangTilEktefelle())
                .withFolkeregistrertMedEktefelle(ektefelle.getFolkeregistrertSammen());
    }

    private static JsonEktefelle tilSystemregistrertJsonEktefelle(Ektefelle ektefelle) {
        if (ektefelle == null || ektefelle.getIkkeTilgangTilEktefelle()) {
            return new JsonEktefelle().withNavn(new JsonNavn()
                    .withFornavn("")
                    .withMellomnavn("")
                    .withEtternavn(""));
        }
        return new JsonEktefelle()
                .withNavn(mapToJsonNavn(ektefelle))
                .withFodselsdato(ektefelle.getFodselsdato() != null ? ektefelle.getFodselsdato().toString() : null)
                .withPersonIdentifikator(ektefelle.getFnr());
    }

    private static JsonNavn mapToJsonNavn(Ektefelle ektefelle) {
        return new JsonNavn()
                .withFornavn(ektefelle.getFornavn() != null ? ektefelle.getFornavn() : "")
                .withMellomnavn(ektefelle.getMellomnavn() != null ? ektefelle.getMellomnavn() : "")
                .withEtternavn(ektefelle.getEtternavn() != null ? ektefelle.getEtternavn() : "");
    }

    private JsonForsorgerplikt innhentSystemverdiForsorgerplikt(String personIdentifikator) {
        JsonForsorgerplikt jsonForsorgerplikt = new JsonForsorgerplikt().withHarForsorgerplikt(new JsonHarForsorgerplikt());

        var barn = personService.hentBarnForPerson(personIdentifikator);
        if (barn == null || barn.isEmpty()) {
            jsonForsorgerplikt.getHarForsorgerplikt()
                    .withKilde(JsonKilde.SYSTEM)
                    .withVerdi(false);
            return jsonForsorgerplikt;
        }
        jsonForsorgerplikt.getHarForsorgerplikt()
                .withKilde(JsonKilde.SYSTEM)
                .withVerdi(true);
        jsonForsorgerplikt.setAnsvar(barn.stream().map(this::mapToJsonAnsvar)
                .collect(Collectors.toList()));
        return jsonForsorgerplikt;
    }

    private JsonAnsvar mapToJsonAnsvar(Barn barn) {
        return new JsonAnsvar()
                .withBarn(new JsonBarn()
                        .withKilde(JsonKilde.SYSTEM)
                        .withNavn(new JsonNavn()
                                .withFornavn(barn.getFornavn())
                                .withMellomnavn(barn.getMellomnavn())
                                .withEtternavn(barn.getEtternavn()))
                        .withFodselsdato(barn.getFodselsdato() != null ? barn.getFodselsdato().toString() : null)
                        .withPersonIdentifikator(barn.getFnr())
                        .withHarDiskresjonskode(false))
                .withErFolkeregistrertSammen(new JsonErFolkeregistrertSammen()
                        .withKilde(JsonKildeSystem.SYSTEM)
                        .withVerdi(barn.getFolkeregistrertSammen()));
    }

}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/personalia/familie/FamilieSystemdata.kt
package no.nav.sosialhjelp.soknad.personalia.familie

import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKildeSystem
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonNavn
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonAnsvar
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonBarn
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonEktefelle
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonErFolkeregistrertSammen
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonForsorgerplikt
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonHarForsorgerplikt
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonSivilstatus
import no.nav.sosialhjelp.soknad.app.systemdata.Systemdata
import no.nav.sosialhjelp.soknad.personalia.person.PersonService
import no.nav.sosialhjelp.soknad.personalia.person.domain.Barn
import no.nav.sosialhjelp.soknad.personalia.person.domain.Ektefelle
import no.nav.sosialhjelp.soknad.repository.soknadunderarbeid.SoknadUnderArbeid
import org.apache.commons.lang3.StringUtils.isEmpty
import org.springframework.stereotype.Component

@Component
class FamilieSystemdata(
    private val personService: PersonService
) : Systemdata {

    override fun updateSystemdataIn(soknadUnderArbeid: SoknadUnderArbeid) {
        val jsonData = soknadUnderArbeid.jsonInternalSoknad?.soknad?.data ?: return
        val personIdentifikator = jsonData.personalia.personIdentifikator.verdi
        val familie = jsonData.familie
        val systemverdiSivilstatus = innhentSystemverdiSivilstatus(personIdentifikator)

        if (systemverdiSivilstatus != null || familie.sivilstatus == null || familie.sivilstatus.kilde == JsonKilde.SYSTEM) {
            familie.sivilstatus = systemverdiSivilstatus
        }

        val forsorgerplikt = familie.forsorgerplikt
        val harForsorgerplikt = forsorgerplikt.harForsorgerplikt
        val systemverdiForsorgerplikt = innhentSystemverdiForsorgerplikt(personIdentifikator)

        if (systemverdiForsorgerplikt.harForsorgerplikt.verdi) {
            forsorgerplikt.harForsorgerplikt = systemverdiForsorgerplikt.harForsorgerplikt

            val ansvarList = forsorgerplikt.ansvar?.toMutableList()
            if (!ansvarList.isNullOrEmpty()) {
                ansvarList.removeIf { it.barn.kilde == JsonKilde.SYSTEM && isNotInList(it, systemverdiForsorgerplikt.ansvar) }
                ansvarList.addAll(
                    systemverdiForsorgerplikt.ansvar.filter { isNotInList(it, forsorgerplikt.ansvar) }
                )
            } else {
                forsorgerplikt.ansvar = systemverdiForsorgerplikt.ansvar
            }
        } else if (harForsorgerplikt == null || harForsorgerplikt.kilde == JsonKilde.SYSTEM || !harForsorgerplikt.verdi) {
            forsorgerplikt.harForsorgerplikt = systemverdiForsorgerplikt.harForsorgerplikt
            forsorgerplikt.barnebidrag = null
            forsorgerplikt.ansvar = ArrayList()
        }
    }

    private fun isNotInList(jsonAnsvar: JsonAnsvar, jsonAnsvarList: List<JsonAnsvar>): Boolean {
        return jsonAnsvarList.none {
            checkNotNull(it.barn) { "JsonAnsvar mangler barn. Ikke mulig å skille fra andre barn" }
            return if (it.barn.personIdentifikator != null) {
                it.barn.personIdentifikator == jsonAnsvar.barn.personIdentifikator
            } else {
                it.barn.navn == jsonAnsvar.barn.navn
            }
        }
    }

    private fun innhentSystemverdiSivilstatus(personIdentifikator: String): JsonSivilstatus? {
        val person = personService.hentPerson(personIdentifikator)
        if (person == null || isEmpty(person.sivilstatus)) {
            return null
        }
        val ektefelle = person.ektefelle
        val status = JsonSivilstatus.Status.fromValue(person.sivilstatus)
        return if (JsonSivilstatus.Status.GIFT != status || ektefelle == null) {
            null
        } else JsonSivilstatus()
            .withKilde(JsonKilde.SYSTEM)
            .withStatus(status)
            .withEktefelle(tilSystemregistrertJsonEktefelle(ektefelle))
            .withEktefelleHarDiskresjonskode(ektefelle.ikkeTilgangTilEktefelle)
            .withFolkeregistrertMedEktefelle(ektefelle.folkeregistrertSammen)
    }

    private fun innhentSystemverdiForsorgerplikt(personIdentifikator: String): JsonForsorgerplikt {
        val jsonForsorgerplikt = JsonForsorgerplikt().withHarForsorgerplikt(JsonHarForsorgerplikt())
        val barn = personService.hentBarnForPerson(personIdentifikator)
        if (barn.isNullOrEmpty()) {
            jsonForsorgerplikt.harForsorgerplikt
                .withKilde(JsonKilde.SYSTEM)
                .withVerdi(false)
            return jsonForsorgerplikt
        }
        jsonForsorgerplikt.harForsorgerplikt
            .withKilde(JsonKilde.SYSTEM)
            .withVerdi(true)
        jsonForsorgerplikt.ansvar = barn.map { mapToJsonAnsvar(it) }
        return jsonForsorgerplikt
    }

    private fun mapToJsonAnsvar(barn: Barn): JsonAnsvar {
        return JsonAnsvar()
            .withBarn(
                JsonBarn()
                    .withKilde(JsonKilde.SYSTEM)
                    .withNavn(
                        JsonNavn()
                            .withFornavn(barn.fornavn)
                            .withMellomnavn(barn.mellomnavn)
                            .withEtternavn(barn.etternavn)
                    )
                    .withFodselsdato(barn.fodselsdato?.toString())
                    .withPersonIdentifikator(barn.fnr)
                    .withHarDiskresjonskode(false)
            )
            .withErFolkeregistrertSammen(
                JsonErFolkeregistrertSammen()
                    .withKilde(JsonKildeSystem.SYSTEM)
                    .withVerdi(barn.folkeregistrertSammen)
            )
    }

    companion object {
        private fun tilSystemregistrertJsonEktefelle(ektefelle: Ektefelle?): JsonEktefelle {
            return if (ektefelle == null || ektefelle.ikkeTilgangTilEktefelle) {
                JsonEktefelle().withNavn(
                    JsonNavn()
                        .withFornavn("")
                        .withMellomnavn("")
                        .withEtternavn("")
                )
            } else JsonEktefelle()
                .withNavn(mapToJsonNavn(ektefelle))
                .withFodselsdato(ektefelle.fodselsdato?.toString())
                .withPersonIdentifikator(ektefelle.fnr)
        }

        private fun mapToJsonNavn(ektefelle: Ektefelle): JsonNavn {
            return JsonNavn()
                .withFornavn(ektefelle.fornavn ?: "")
                .withMellomnavn(ektefelle.mellomnavn ?: "")
                .withEtternavn(ektefelle.etternavn ?: "")
        }
    }
}
